packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,219 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {camelCase} from 'lodash';
+import {DataObject} from '../../common-types';
+import {EntityNotFoundError, InvalidRelationError} from '../../errors';
+import {Entity} from '../../model';
+import {EntityCrudRepository} from '../../repositories/repository';
+import {constrainFilter} from '../../repositories/constraint-utils';
+import {isTypeResolver} from '../../type-resolver';
+import {Getter, HasManyThroughDefinition} from '../relation.types';
+import {
+  DefaultHasManyThroughRepository,
+  HasManyThroughRepository,
+} from './has-many-through.repository';
+import {Filter, Where} from '../../query';
+
+const debug = debugFactory('loopback:repository:has-many-repository-factory');
+
+export type HasManyThroughRepositoryFactory<
+  Target extends Entity,
+  Through extends Entity,
+  ForeignKeyType
+> = (fkValue: ForeignKeyType) => HasManyThroughRepository<Target, Through>;
+
+export interface AdvancedConstraint<T extends Entity> {
+  dataObject: DataObject<T>;
+  filter: Filter<T>;
+  where: Where<T>;
+}
+
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. For example, if a Customer model is related to an Order model
+ * via a HasMany relation, then, the relational repository returned by the
+ * factory function would be constrained by a Customer model instance's id(s).
+ *
+ * @param relationMetadata The relation metadata used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepositoryGetter The repository which represents the target model of a
+ * relation attached to a datasource.
+ * @returns The factory function which accepts a foreign key value to constrain
+ * the given target repository
+ */
+export function createHasManyThroughRepositoryFactory<
+  Target extends Entity,
+  TargetID,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  relationMetadata: HasManyThroughDefinition,
+  targetRepositoryGetter: Getter<EntityCrudRepository<Target, TargetID>>,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+): HasManyThroughRepositoryFactory<Target, Through, ForeignKeyType> {
+  const meta = resolveHasManyThroughMetadata(relationMetadata);
+  debug('Resolved HasMany relation metadata: %o', meta);
+  return function(fkValue?: ForeignKeyType) {
+    async function getAdvancedConstraint(
+      targetInstance: Target,
+    ): Promise<AdvancedConstraint<Target | Through>> {
+      return createAdvancedConstraint<
```

Comment:
Similarly here, how about `createHasManyThroughConstraint`?

Reply:
Refer to the previous comment

Reply:
I have separated the constraint creator into `createTargetConstraint` and `createThroughConstraint`. Read the following comment.

https://github.com/strongloop/loopback-next/pull/2359#issuecomment-475979639